### PR TITLE
OL-2555 and OL-2556 email_Send Kicoff and Swagger file

### PR DESCRIPTION
**Pull request by narayanangp**
-  New service kick off - NSI_OL052_TLS02_SEND_EMAIL
-  Updating  Swagger File as per the API Document

### CODE REVIEW

File: swagger/sendEmail.yaml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        </table>
+        </br>
+
+
+      type: object
+      properties:
+        errorCode:
+          description: Error code identifying cause of the problem, format E####
+          type: string
+          example: EAUT001
```
Comment (jnyckowiak) [inline swagger/sendEmail.yaml]: please use code relevant to the API e.g. E0001

Comment (narayanangp) [reply swagger/sendEmail.yaml]: fixed

File: swagger/sendEmail.yaml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        folderId:
+          type: integer
+          description: Number of the folder containg the triggered emails(s)
+          example: 12
+        emailAddress:
+          type: string
+          description: Email address of the recipient
+          example: test@atos.com
+        profileData:
+          type: array
```
Comment (pkaliniak) [inline swagger/sendEmail.yaml]: This one looks suspicous to me. Example for XP has following sample request:
```
profileData={  
   "TFC":{  
      "RecipientName":"Matthew",
      "ProductShortName":"Childcareservice",
      "ProductName":"Childcare service"
   }
}
``` 
That is not an array. I'll check with Mike

**Comment by narayanangp** [comment]
> Did you use job for Service generation ? It looks like job for stub generatoin was run here.
> More comments below.

No, Archetype for rest was not usable and it was disable and deleted, Created a new service from the existing one (OL046)

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+FROM spring-boot-fuse-ol:1.0-SNAPSHOT
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply Dockerfile]: Just changed, please check

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-configuration-processor</artifactId>
+            <optional>true</optional>
+        </dependency>
+
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <version>${rest-assured.version}</version>
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply pom.xml]: removed

File: src/main/resources/logback.xml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<configuration>
+
+    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply src/main/resources/logback.xml]: this is done.

File: swagger/sendEmail.yaml (hunk 2 above)
Comment (narayanangp) [reply swagger/sendEmail.yaml]: Please check and let me know

File: swagger/sendEmail.yaml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      requestBody:
+        description: Email send request body
+        required: true
+        content:
+          application/x-www-form-urlencoded:
+            schema:
+              $ref: '#/components/schemas/Request'
+
+      responses:
+        '200':
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply swagger/sendEmail.yaml]: Sure, Please check and let me know

[the remark this answers is not in the document]
Comment (pkaliniak) [reply swagger/sendEmail.yaml]: That change is approved

[the remark this answers is not in the document]
Comment (narayanangp) [reply swagger/sendEmail.yaml]: I changed the response from 200 to 204. Please check.

File: swagger/sendEmail.yaml (hunk 2 above)
Comment (narayanangp) [reply swagger/sendEmail.yaml]: I have made the changes, plz review.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,189 @@
...
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
+``````
+
+Response:
+Http code 204
+```javascript
+{
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply README.md]: fixed

File: swagger/sendEmail.yaml [hunk 8]
```diff
@@ -0,0 +1,157 @@
+openapi: 3.0.0
+info:
+  version: "1.0"
+  title: "NS&I Orchestration Layer - Send Email"
+  description: "Swagger definition"
+paths:
+  "/messaging/v1/email/email_send":
```
[the remark this answers is not in the document]
Comment (narayanangp) [reply swagger/sendEmail.yaml]: fixed

File: swagger/sendEmail.yaml [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+          in: header
+          schema:
+            type: string
+          required: true
+
+      requestBody:
+        description: Email send request body
+        required: true
+        content:
+          application/x-www-form-urlencoded:
```
Comment (pkaliniak) [inline swagger/sendEmail.yaml]: nope, we don't need to use this contetnt type. When you use it then all the parameters are logged as they are.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,184 @@
...
+```bash
+docker inspect -f '{{.NetworkSettings.Networks.fuse.IPAddress}}' ol052-tls02-send-email
+```
+
+### Case examples:
+1) Try to send an email with a valid request
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: * Change content type (and reflect that in swagger) 
* change path to reflect swagger
* change request, we don't send `TFC` in request,

File: README.md [hunk 11]
```diff
@@ -1,2 +1,184 @@
...
+``````
+
+Response:
+Http code 204
+
+2) Try to send an email with the no body
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 12]
```diff
@@ -1,2 +1,184 @@
...
+    "errorCode": "E0000"
+    "message": "Invalid input data"
+}
+```
+
+3) Try to send an email, when 'apiRequestOriginator' header is missing
+
+Request:
+```bash
+curl -v -XPOST -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 13]
```diff
@@ -1,2 +1,184 @@
...
+    "errorCode": "E0001"
+    "message": "apiRequestOriginator is missing"
+}
+```
+
+4) Try to send an email, when 'RequestID' header is missing
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 14]
```diff
@@ -1,2 +1,184 @@
...
+}
+```
+
+5) Try to send email, when returning 403 error during accessing the Service Layer
+
+The trick for test, this call is to have active service and stub but with no authentication
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 15]
```diff
@@ -1,2 +1,184 @@
...
+}
+```
+
+6) Try to send email, when returning 503 error during accessing the Service Layer
+
+The trick for test, this call is to have active service, but deactivated stub (so the call from service would not be able to connect with stub)
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 16]
```diff
@@ -1,2 +1,184 @@
...
+}
+```
+
+7) Try to send email, when returning 503 error during accessing the Service Layer
+
+The trick for test, this call is to have active service, but stub should not respond (due to technical issue) eventhough stub is up and running
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 17]
```diff
@@ -1,2 +1,184 @@
...
+Response:
+Http code 503
+```javascript
+{
+    "errorCode": "E8001",
+    "errorDescription": "Service not available due to lack of connection to provider"
+}
+```
+
+8) Simulate receiving Service Layer Internal Server Error
```
Comment (pkaliniak) [inline README.md]: as above

File: README.md [hunk 18]
```diff
@@ -1,2 +1,184 @@
...
+    "errorCode": "E9000",
+    "errorDescription": "Internal server error"
+}
+```
+
+9) Simulate receiving Service Layer Internal Server Error, due to the provider returns an error condition that was not mapped in the table
+
+Request:
+```bash
+curl -v -XPOST -H "apiRequestOriginator:123" -H "RequestID:6948DF80-14BD-4E04-8842-7668D9C001F5" "Content-Type: application/x-www-form-urlencoded" -d 'http://127.0.0.1:1080/api/json/email_send/method=triggerFolder&folderId=47&emailAddress=matthew.2.johnson@atos.net&profileData={"TFC":{"RecipientName":"Matthew","ProductShortName":"Childcareservice","ProductName":"Childcare service"}}'
```
Comment (pkaliniak) [inline README.md]: as above

File: pom.xml [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+
+    <name>l052-tls02-send-email - Spring-boot with fuse and camel</name>
+    <description>Spring Boot with fuse</description>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.java.version>1.8</project.java.version>
+
+        <!-- Core dependency versions -->
+        <fuse-eap.version>7.2.0.fuse-720020-redhat-00001</fuse-eap.version>
```
Comment (pkaliniak) [inline pom.xml]: we don't have support for 7.x only for 6.x. We will migrate when we switch to openshift

File: pom.xml [hunk 20]
```diff
@@ -0,0 +1,405 @@
...
+        <!-- configure the versions you want to use here -->
+        <fabric8.version>2.2.170.redhat-000023</fabric8.version>
+        <!--same version as in parent POM: version.org.springframework.boot.spring-boot-->
+
+
+        <!-- maven plugin versions -->
+        <maven-compiler-plugin.version>3.3</maven-compiler-plugin.version>
+        <maven-surefire-plugin.version>2.18.1</maven-surefire-plugin.version>
+        <version.jacoco>0.7.8</version.jacoco>
+        <org.mapstruct.version>1.2.0.Final</org.mapstruct.version>
```
Comment (pkaliniak) [inline pom.xml]: why do you need that?

File: pom.xml [hunk 21]
```diff
@@ -0,0 +1,405 @@
...
+            <!-- needs to be higher than 1.9.2 to use newer docker -->
+            <version>${arquillian-cube-docker.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.github.docker-java</groupId>
+            <artifactId>docker-java</artifactId>
+            <version>${version.github.docker-java}</version>
+        </dependency>
```
Comment (pkaliniak) [inline pom.xml]: scope `provided`

File: pom.xml (hunk 19 above)
Comment (narayanangp) [reply pom.xml]: I checked with other latest services like 049, 050 and 051 everyone uses the same fuse version
<fuse-eap.version>7.2.0.fuse-720020-redhat-00001</fuse-eap.version>

Do you want me to change it to 6.0? Do you have the correct 6.x version?

File: pom.xml (hunk 20 above)
Comment (narayanangp) [reply pom.xml]: This is was my first kickoff, so was trying to copy from other services, Just removed now.

**Pull request merged**